Add BROS

jinhopark8345 opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Add [BROS(BERT Relying On Spatiality)](https://arxiv.org/abs/2108.04539) to 🤗 Transformers


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a [Github issue](https://github.com/huggingface/transformers/issues/23181) or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@NielsRogge

jinhopark8345 committed:
add Bros boilerplate

jinhopark8345 committed:
copy and pasted modeling_bros.py from official Bros repo

jinhopark8345 committed:
update copyright of bros files

jinhopark8345 committed:
copy tokenization_bros.py from official repo and update import path

jinhopark8345 committed:
copy tokenization_bros_fast.py from official repo and update import path

jinhopark8345 committed:
copy configuration_bros.py from official repo and update import path

jinhopark8345 committed:
remove trailing period in copyright line

jinhopark8345 committed:
copy and paste bros/__init__.py from official repo

jinhopark8345 committed:
save formatting

jinhopark8345 committed:
remove unused unnecessary pe_type argument - using only crel type

jinhopark8345 commented:
@amyeroberts 

I am confused about what needs to be done.

According to the [How to add a new model](https://huggingface.co/docs/transformers/add_new_model#514-port-brandnewbert-to-transformers) guideline, a big part of it is porting pretrained models (from the original repo) into Huggingface transformers and making sure they are correctly ported by checking the outputs of each layer's forward step.

However, it seems like the authors of the Bros model used `transformers-cli` to create the boilerplate code, and I don't think there is much to change from the original code.

Do I need to write a conversion script? Or can I skip this step?

jinhopark8345 committed:
resolve import issue

jinhopark8345 committed:
remove unused model classes

jinhopark8345 committed:
remove unnecessary tests

jinhopark8345 committed:
remove unused classes

jinhopark8345 committed:
fix original code's bug - layer_module's argument order

jinhopark8345 committed:
clean up modeling auto

jinhopark8345 committed:
add bbox to prepare_config_and_inputs

jinhopark8345 committed:
set temporary value to hidden_size (32 is too low because of the of the
Bros' positional embedding)

jinhopark8345 committed:
remove decoder test, update create_and_check* input arguemnts

jinhopark8345 committed:
add missing variable to model tests

jinhopark8345 committed:
do make fixup

jinhopark8345 committed:
update bros.mdx

jinhopark8345 committed:
add boilerate plate for no_head inference test

jinhopark8345 committed:
update BROS_PRETRAINED_MODEL_ARCHIVE_LIST (add naver-clova-ocr prefix)

jinhopark8345 committed:
add prepare_bros_batch_inputs function

jinhopark8345 committed:
add prepare_bros_batch_inputs function

jinhopark8345 committed:
update modeling_common to add bbox inputs in Bros Model Test

jinhopark8345 committed:
remove unnecessary model inference

jinhopark8345 committed:
add test case

jinhopark8345 committed:
add test case

jinhopark8345 committed:
add model_doc

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_23190). All of your documentation changes will be reflected on that endpoint.

jinhopark8345 committed:
add test case for token_classification

jinhopark8345 committed:
add test case for token_classification

jinhopark8345 committed:
apply fixup

jinhopark8345 commented:
Still working on it! Writing tutorial/demo notebooks of how to use BROS

jinhopark8345 committed:
update modeling code

jinhopark8345 committed:
update modeling code

jinhopark8345 commented:
@amyeroberts Is it possible to reopen this PR because I have been working on [forked repo](https://github.com/jinhopark8345/bros/tree/feature/update-data-loading)  of original bros code

jinhopark8345 committed:
update BrosForTokenClassification loss calculation logic

jinhopark8345 committed:
update BrosForTokenClassification loss calculation logic

jinhopark8345 committed:
revert logits preprocessing logic to make sure logits have original shape

jinhopark8345 committed:
revert logits preprocessing logic to make sure logits have original shape

jinhopark8345 committed:
- update class name

jinhopark8345 committed:
- add BrosSpadeOutput
- update BrosConfig arguments

jinhopark8345 committed:
add boilerate plate for no_head inference test

jinhopark8345 committed:
Merge remote-tracking branch 'upstream/main' into add-bros

jinhopark8345 committed:
apply masking on the fly

jinhopark8345 committed:
add BrosSpadeForTokenLinking

jinhopark8345 committed:
update class name
put docstring to the beginning of the file

jinhopark8345 committed:
separate the logits calculation logic and loss calculation logic

jinhopark8345 committed:
update logic for loss calculation so that logits shape doesn't change
when return

jinhopark8345 committed:
update typo

jinhopark8345 committed:
update prepare_config_and_inputs

jinhopark8345 committed:
update dummy node initialization

jinhopark8345 committed:
update last_hidden_states getting logic to consider when return_dict is False

jinhopark8345 committed:
update box first token mask param

jinhopark8345 committed:
bugfix: remove random attention mask generation

jinhopark8345 committed:
update keys to ignore on load missing

jinhopark8345 committed:
run make style and quality

jinhopark8345 committed:
apply make style and quality of other codes

jinhopark8345 committed:
update box_first_token_mask to bool type

jinhopark8345 committed:
update index.md

jinhopark8345 committed:
apply make style and quality

jinhopark8345 committed:
Merge remote-tracking branch 'upstream/main' into add-bros

jinhopark8345 committed:
apply make fix-copies

jinhopark8345 committed:
pass check_repo

jinhopark8345 committed:
update bros model doc

jinhopark8345 committed:
docstring bugfix fix

jinhopark8345 committed:
add checkpoint for doc, tokenizer for doc

amyeroberts commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+            self_token_mask = torch.eye(max_seq_length, max_seq_length + 1).to(device).bool()
+
+            mask = box_first_token_mask.view(-1)
+            box_first_token_mask = torch.cat(
+                [
+                    ~box_first_token_mask,
+                    torch.zeros([bsz, 1], dtype=torch.bool).to(device),
+                ],
+                axis=1,
+            )
```
```suggestion
            box_first_token_mask = torch.cat(
                [~box_first_token_mask, torch.zeros([batch_size, 1], dtype=torch.bool, device=device)], axis=1,
            )
```

jinhopark8345 replied:
Apply suggestions from code review.

amyeroberts commented on docs/source/en/model_doc/bros.md:
````diff
@@ -0,0 +1,84 @@
+    bboxes[:, [1, 3, 5, 7]] = bboxes[:, [1, 3, 5, 7]] / doc_height
+```
+
+- [`~transformers.BrosForTokenClassification.forward`, `~transformers.BrosSpadeEEForTokenClassification.forward`, `~transformers.BrosSpadeEEForTokenClassification.forward`] require not only *input_ids* and *bbox* but also `box_first_token_mask` for loss calculation. It is a mask to filter out non-first tokens of each box. You can obtain this mask by saving start token indices of bounding boxes wh ...
````
It would be more useful to provide an example directly here

jinhopark8345 replied:
Added an example!

amyeroberts reviewed (commented):
Thanks for adding this model! 

Overall PR is looking good - very clean and easy to read. Most comments are nits about formatting and have the copied from statements. 

What's the difference between Bros, BrosSpadeEL and BrosSpadeEE ? The purpose of these three modules isn't clear. Could you add an explanation in the model's class docstrings as well as the model's doc page - bros.md? 

About the model inputs. We try as much as possible in transformers to have a consistent API. This also means consistent inputs and outputs. It would be better if this model took a standard bbox format e.g. (x0, y0, x1, y1) and then converted to the 6 point needed within its forward pass.

jinhopark8345 committed:
Update README.md

jinhopark8345 committed:
Update docs/source/en/model_doc/bros.md

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

jinhopark8345 committed:
Update bros.md

jinhopark8345 committed:
Update src/transformers/__init__.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

jinhopark8345 committed:
Update docs/source/en/model_doc/bros.md

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

jinhopark8345 committed:
Apply suggestions from code review

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

jinhopark8345 committed:
apply suggestions from code review

jinhopark8345 committed:
apply suggestions from code review

jinhopark8345 committed:
revert test_processor_markuplm.py

jinhopark8345 committed:
Update test_processor_markuplm.py

jinhopark8345 committed:
apply suggestions from code review

jinhopark8345 commented:
> Thanks for adding this model!
> 
> Overall PR is looking good - very clean and easy to read. Most comments are nits about formatting and have the copied from statements.
> 

Thank you for the feedback!!

> What's the difference between Bros, BrosSpadeEL and BrosSpadeEE ? The purpose of these three modules isn't clear. Could you add an explanation in the model's class docstrings as well as the model's doc page - bros.md?

Bros is a encoder transformer. BrosSpadeEL is Bros + [SPADE EL](https://arxiv.org/abs/2005.00642) (entity linking decoder attached) and BrosSpadeEE is Bros +  [SPADE EE](https://arxiv.org/abs/2005.00642) (entity extraction decoder attached). I will add more detailed explanation to model's doc page.

[Original bros repo](https://github.com/clovaai/bros) has total  3 heads for the Bros Model:
- BrosTokenClassification
- BrosEE (entity extraction)
- BrosEL (entity linking)

Both BrosTokenClassification and BrosEE essentially perform the same job. However, in the case of BrosTokenClassification, it assumes input tokens are perfectly serialized (which is very challenging task since they exist in a 2D space), while BrosEE allows for more flexibility in handling serialization errors as it predicts next connection token from one token.

BrosEL predicts relation from first token (of one entity) to the other first token(of the other entity) if these two entities share some relation.


> About the model inputs. We try as much as possible in transformers to have a consistent API. This also means consistent inputs and outputs. It would be better if this model took a standard bbox format e.g. (x0, y0, x1, y1) and then converted to the 6 point needed within its forward pass.

I will change the forward pass to convert from 2 points to 4 points within its forward pass.

jinhopark8345 committed:
apply suggestions from code review

jinhopark8345 committed:
apply suggestions from code review

jinhopark8345 committed:
update BrosSpadeELForTokenClassification head name to entity linker

jinhopark8345 committed:
add doc string for config params

jinhopark8345 commented on src/transformers/models/bros/configuration_bros.py:
```diff
@@ -0,0 +1,134 @@
+        initializer_range=0.02,
+        layer_norm_eps=1e-12,
+        pad_token_id=0,
+        bbox_scale=100.0,
+        n_relations=1,
```
Updated doc string

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+        return bbox_pos_emb
+
+
+class BrosEmbeddings(nn.Module):
```
`BrosEmbedding` has `bbox_sinusoid_emb` and `bbox_projection` layer for 2D embedding that `BertEmeddings` doesn't have. But except those two, they are quite similiar as far as I know, Would it still be valid to put the "Copied ~" comment?

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+        embeddings = self.dropout(embeddings)
+        return embeddings
+
+    def calc_bbox_pos_emb(self, bbox):
```
I am little confused with what you meant by module level function.
Do you mean that it would be better to divide the forward method into multiple methods?

jinhopark8345 committed:
update class, var names to more explicit and apply suggestions from code review

jinhopark8345 committed:
remove unnecessary keys to ignore

jinhopark8345 committed:
update relation extractor to be initialized with config

jinhopark8345 committed:
add bros processor

jinhopark8345 committed:
apply make style and quality

jinhopark8345 committed:
update bros.md

jinhopark8345 committed:
remove bros tokenizer, add bros processor that wraps bert tokenizer

jinhopark8345 committed:
revert change

jinhopark8345 committed:
apply make fix-copies

jinhopark8345 committed:
update processor code, update itc -> initial token, stc -> subsequent token

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+        input_ids=None,
+        bbox=None,
+        attention_mask=None,
+        box_first_token_mask=None,
```
updated `BROS_INPUTS_DOCSTRING`

jinhopark8345 committed:
add type hint

jinhopark8345 committed:
remove unnecessary condition branches in embedding forward

jinhopark8345 committed:
Merge remote-tracking branch 'upstream/main' into add-bros

jinhopark8345 committed:
fix auto tokenizer fail

jinhopark8345 commented:
@amyeroberts 
Would it be a better idea to add/update `tokenizer_bros.py `and `processing_bros.py` that takes boxes as input, similar to [layoutlmv3](https://huggingface.co/docs/transformers/model_doc/layoutlmv3) ?

while  updating the doctest code for the bros model, I realized that trying out the bros model using ([my demo script](https://github.com/jinhopark8345/FormUnderstanding/blob/main/scripts/bros_on_SROIE_BIO_tagging.py)) would be quite complicated.

jinhopark8345 committed:
update docstring for each classes

jinhopark8345 committed:
Merge remote-tracking branch 'upstream/main' into add-bros

amyeroberts commented:
@jinhopark8345 What I would suggest is following the pattern of LayoutLM model: the model takes in bboxes of format (x0, y0, x1, y1). Then for BROs, within the model these are converted to the 6 points used in the forward pass.

No tokenizer needs to be added as the model can use the already implemented BERTTokenizer. 

Processors classes which bundle together two processing classes e.g. tokenizer + image processor, so isn't needed here.

jinhopark8345 committed:
update bbox input dimension as standard 2 points and convert them to 4
points in forward pass

jinhopark8345 commented:
@amyeroberts I am encountering the following errors during the tests ( `ci/circleci : tests_torch` )

```bash
FAILED tests/utils/test_hub_utils.py::GetFromCacheTests::test_get_file_gated_repo - AssertionError: OSError not raised
FAILED tests/utils/test_hub_utils.py::GetFromCacheTests::test_has_file_gated_repo - AssertionError: OSError not raised
```
Is this related to my pull request, or are there issues within the Huggingface library?

amyeroberts commented:
@jinhopark8345  - these aren't related to your PR and are likely transient issues with connecting to the hub. You can ignore them for now. If they persist exclusively on this PR and when the PR is finalised then we can dig into it more.

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+        return bbox_pos_emb
+
+
+class BrosSelfAttention(nn.Module):
```
Same as above, 
```python
    self.bbox_sinusoid_emb = BrosPositionalEmbedding2D(config)
    self.bbox_projection = nn.Linear(config.dim_bbox_sinusoid_emb_2d, config.dim_bbox_projection, bias=False)
```
is the difference in `__init__` method compared to 
Would it still be valid to put the "Copied ~" comment?

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+        return outputs
+
+
+class BrosSelfOutput(nn.Module):
```
Bros has some differences with `BertEmbedding`. (I assumed that you are suggesting adding Copied ~ statement )

Bros calculate attention scores by combing both semantic and spatial features (line 390), and the `query_layer` is shared when calculating attention scores from semantic features and spatial features. (line 388).

jinhopark8345 replied:
Applied suggestions from code review!

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+        return hidden_states
+
+
+class BrosAttention(nn.Module):
```
Added Copied statements!

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+        return outputs
+
+
+class BrosIntermediate(nn.Module):
```
Here, forward method gets `bbox_pos_emb` but BertAttention's forward method doesn't.

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+        return hidden_states
+
+
+class BrosOutput(nn.Module):
```
Added Copied statements!

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+        return layer_output
+
+
+class BrosEncoder(nn.Module):
```
Here, `BrosLayer`'s forward method gets `bbox_pos_emb` and pass with it to attention layer but `BertLayer` doesn't.

jinhopark8345 committed:
update bros docs

jinhopark8345 commented on docs/source/en/model_doc/bros.md:
```diff
@@ -0,0 +1,84 @@
+
+## Overview
+
+The Bros model was proposed in [BROS: A Pre-trained Language Model Focusing on Text and Layout for Better Key Information Extraction from Documents](https://arxiv.org/abs/2108.04539)  by Teakgyu Hong, Donghyun Kim, Mingi Ji, Wonseok Hwang, Daehyun Nam, Sungrae Park. BROS is a document understanding model pre-trained with the area-masking strategy. It obtains comparable or better result on KIE ben ...
```
Added a simple description.

jinhopark8345 commented on docs/source/en/model_doc/bros.md:
```diff
@@ -0,0 +1,84 @@
+
+Tips:
+
+- [`~transformers.BrosModel.forward`] requires *input_ids* and `bbox` (bounding box). Each bounding box should be in (x0, y0, x1, y0, x1, y1, x0, y1) format represented by four clockwise points starting from top-left corner. Obtaining of Bounding boxes depends on external OCR system. The `x` coordinate should be normalized by document image width, and the `y` coordinate should be normalized by do ...
```
Applied suggestions from code review.

jinhopark8345 commented on src/transformers/__init__.py:
```diff
@@ -4946,8 +4961,6 @@
         )
         from .modeling_utils import PreTrainedModel
-
-        # PyTorch model imports
```
Apply suggestions from code review.

jinhopark8345 commented on src/transformers/__init__.py:
```diff
@@ -5193,6 +5206,16 @@
             BridgeTowerPreTrainedModel,
         )
+
+        # PyTorch model imports
```
Apply suggestions from code review.

jinhopark8345 commented on src/transformers/modeling_outputs.py:
```diff
@@ -1660,3 +1660,33 @@ def logits(self):
+    Base class for outputs of token classification models.
+
+    Args:
+        loss (`torch.FloatTensor` of shape `(1,)`, *optional*, returned when `labels` is provided) :
```
Apply suggestions from code review.

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+
+        head_mask (:
+            obj:*torch.FloatTensor* of shape `(num_heads,)` or `(num_layers, num_heads)`, *optional*): Mask to nullify
+            selected heads of the self-attention modules. Mask values selected in `[0, 1]`:
```
Apply suggestions from code review.

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+"""
+
+
+class PositionalEmbedding1D(nn.Module):
```
Apply suggestions from code review.

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+class PositionalEmbedding1D(nn.Module):
+    # Reference: https://github.com/kimiyoung/transformer-xl/blob/master/pytorch/mem_transformer.py#L15
+
+    def __init__(self, demb):
```
Apply suggestions from code review.

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+        inv_freq = 1 / (10000 ** (torch.arange(0.0, demb, 2.0) / demb))
+        self.register_buffer("inv_freq", inv_freq)
+
+    def forward(self, pos_seq, bsz=None):
```
Apply suggestions from code review.

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+        if len(seq_size) == 2:
+            b1, b2 = seq_size
+            sinusoid_inp = pos_seq.view(b1, b2, 1) * self.inv_freq.view(1, 1, self.demb // 2)
+        elif len(seq_size) == 3:
```
Removed the if branch.

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+        return pos_emb
+
+
+class PositionalEmbedding2D(nn.Module):
```
Apply suggestions from code review.

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+
+
+class PositionalEmbedding2D(nn.Module):
+    def __init__(self, demb, dim_bbox=8):
```
Apply suggestions from code review.

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+    def __init__(self, config):
+        super().__init__()
+
+        assert config.hidden_size >= 64, "hidden_size of the model must be large or equal to 64"
```
Apply suggestions from code review.

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+        # position_ids (1, len position emb) is contiguous in memory and exported when serialized
+        self.position_embedding_type = getattr(config, "position_embedding_type", "absolute")
+        self.register_buffer("position_ids", torch.arange(config.max_position_embeddings).expand((1, -1)))
+        if version.parse(torch.__version__) > version.parse("1.6.0"):
```
Updated the code.

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+        self.is_decoder = config.is_decoder
+        self.add_cross_attention = config.add_cross_attention
+        if self.add_cross_attention:
+            assert self.is_decoder, f"{self} should be used as a decoder model if cross attention is added"
```
Apply suggestions from code review.

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+
+        cross_attn_present_key_value = None
+        if self.is_decoder and encoder_hidden_states is not None:
+            assert hasattr(
```
Apply suggestions from code review.

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+        return pooled_output
+
+
+class RelationExtractor(nn.Module):
```
Apply suggestions from code review.

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+        backbone_hidden_size,
+        head_hidden_size,
+        head_p_dropout=0.1,
+    ):
```
Apply suggestions from code review.

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+    BROS_START_DOCSTRING,
+)
+class BrosModel(BrosPreTrainedModel):
+    _keys_to_ignore_on_load_missing = [r"embeddings.position_ids"]
```
Removed the line.

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+    )
+    def forward(
+        self,
+        input_ids=None,
```
Added type annotations.

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+        r"""
+        encoder_hidden_states  (:
+            obj:*torch.FloatTensor* of shape `(batch_size, sequence_length, hidden_size)`, *optional*): Sequence of
+            hidden-states at the output of the last layer of the encoder. Used in the cross-attention if the model is
```
Apply suggestions from code review.

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+        past_key_values (:
+            obj:*tuple(tuple(torch.FloatTensor))* of length `config.n_layers` with each tuple having 4 tensors of shape
+            `(batch_size, num_heads, sequence_length - 1, embed_size_per_head)`): Contains precomputed key and value
+            hidden states of the attention blocks. Can be used to speed up decoding.
```
Apply suggestions from code review.

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+                encoder_batch_size,
+                encoder_sequence_length,
+                _,
+            ) = encoder_hidden_states.size()
```
Apply suggestions from code review.

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+        )
+
+        scaled_bbox = bbox * self.config.bbox_scale
+        bbox_pos_emb = self.embeddings.calc_bbox_pos_emb(scaled_bbox)
```
Apply suggestions from code review.

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+    )
+    def forward(
+        self,
+        input_ids=None,
```
Added type annotations.

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+        )
+
+        # Initial token classification for Entity Extraction (NER)
+        self.itc_layer = nn.Sequential(
```
Apply suggestions from code review.

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+        )
+
+        # Subsequent token classification for Entity Extraction (NER)
+        self.stc_layer = RelationExtractor(
```
Apply suggestions from code review.

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+        head_mask=None,
+        inputs_embeds=None,
+        itc_labels=None,
+        stc_labels=None,
```
Updated docstrings.

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+        last_hidden_states = outputs[0]
+        last_hidden_states = last_hidden_states.transpose(0, 1).contiguous()
+        itc_logits = self.itc_layer(last_hidden_states).transpose(0, 1).contiguous()
+        stc_logits = self.stc_layer(last_hidden_states, last_hidden_states).squeeze(0)
```
Apply suggestions from code review.

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+
+        # make stc(sequence token classification) mask
+        inv_attention_mask = 1 - attention_mask
+        bsz, max_seq_length = inv_attention_mask.shape
```
Apply suggestions from code review.

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+    )
+    def forward(
+        self,
+        input_ids=None,
```
Added type annotations.

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1276 @@
+        if labels is not None:
+            loss_fct = CrossEntropyLoss()
+
+            bsz, max_seq_length = attention_mask.shape
```
Apply suggestions from code review.

jinhopark8345 commented on src/transformers/models/bros/tokenization_bros.py:
```diff
@@ -0,0 +1,111 @@
+    return tokens
+
+
+class BrosTokenizer(BertTokenizer):
```
Apply suggestions from code review.

jinhopark8345 commented on src/transformers/models/bros/tokenization_bros_fast.py:
```diff
@@ -0,0 +1,92 @@
+}
+
+
+class BrosTokenizerFast(BertTokenizerFast):
```
Apply suggestions from code review.

jinhopark8345 commented on tests/models/bros/test_modeling_bros.py:
```diff
@@ -0,0 +1,671 @@
+# coding=utf-8
+# Copyright 2022 The HuggingFace Inc. team. All rights reserved.
```
Apply suggestions from code review.

jinhopark8345 commented on tests/models/bros/test_modeling_bros.py:
```diff
@@ -0,0 +1,671 @@
+
+    def test_multi_gpu_data_parallel_forward(self):
+        pass
+
```
Updated code to not just pass.

jinhopark8345 commented on tests/models/markuplm/test_processor_markuplm.py:
```diff
@@ -340,7 +340,7 @@ def test_processor_case_3(self):
             # verify xpath_tags_seq
             # fmt: off
-            expected_xpaths_tags_seq = [[216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216], [109, 25, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 21 ...
+            expected_xpaths_tags_seq = [[216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216], [109, 25, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 21 ...
```
Apply suggestions from code review.

jinhopark8345 commented:
@amyeroberts Thank you very much for the feedback and help! 😃 I truly appreciate it. I have implemented your suggestions from the code review and updated the code. However, there were a few suggestions that I wasn't sure how to address, so I left comments explaining my thoughts on those points.

jinhopark8345 committed:
Merge remote-tracking branch 'upstream/main' into add-bros

amyeroberts commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1312 @@
+
+        bbox ('torch.FloatTensor' of shape '(batch_size, num_boxes, 8)'):
+            Bounding box coordinates for each token in the input sequence. Each bounding box is a list of eight values
+            (x1, y1, x2, y1, x2, y2, x1, y2), where (x1, y1) is the top left corner, and (x2, y2) is the bottom right
```
This should be updated to reflect that 4 point bboxes are now taken

amyeroberts commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1312 @@
+        batch_size, n_head, seq_length, d_head = query_layer.shape
+        bbox_pos_emb = bbox_pos_emb.view(seq_length, seq_length, batch_size, d_head)
+        bbox_pos_emb = bbox_pos_emb.permute([2, 0, 1, 3])
+        bbox_pos_scores = torch.einsum("bnid,bijd->bnij", (query_layer, bbox_pos_emb))
```
Could you replace the `einsum` with classical matrix operations? Unfortunately they currently cause issues when trying to trace the models.

amyeroberts commented on tests/models/bros/test_modeling_bros.py:
```diff
@@ -0,0 +1,448 @@
+        num_labels=3,
+        num_choices=4,
+        scope=None,
+        range_bbox=1000,
```
What is this used for?

amyeroberts commented on src/transformers/models/auto/modeling_auto.py:
```diff
@@ -859,6 +860,7 @@
         ("big_bird", "BigBirdForTokenClassification"),
         ("biogpt", "BioGptForTokenClassification"),
         ("bloom", "BloomForTokenClassification"),
+        ("bros", "BrosForTokenClassification"),
```
I don't think the model can be added here as it requires bboxes as an input

amyeroberts commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1312 @@
+
+                layer_outputs = torch.utils.checkpoint.checkpoint(
+                    create_custom_forward(layer_module),
+                    hidden_states,
+                    attention_mask,
+                    layer_head_mask,
+                    encoder_hidden_states,
+                    encoder_attention_mask,
+                    bbox_pos_emb=bbox_pos_emb,
```
I don't think you can pass in kwargs like this. The arguments should be in the same order as in the forward method

amyeroberts commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1312 @@
+        input_ids: Optional[torch.Tensor] = None,
+        bbox: Optional[torch.Tensor] = None,
+        attention_mask: Optional[torch.Tensor] = None,
+        box_first_token_mask: Optional[torch.Tensor] = None,
```
```suggestion
        bbox_first_token_mask: Optional[torch.Tensor] = None,
```

amyeroberts reviewed (commented):
Very nice work - thank for iterating! 

Just a few small changes - almost ready to merge 🤗

jinhopark8345 committed:
apply suggestions from code review : update Bros -> BROS in bros.md

jinhopark8345 committed:
1. box prefix var -> bbox
2. update variable names to be more explicit

jinhopark8345 committed:
replace einsum with torch matmul

jinhopark8345 committed:
apply style and quality

jinhopark8345 committed:
remove unused argument

jinhopark8345 committed:
remove unused arguments

jinhopark8345 committed:
update docstrings

jinhopark8345 committed:
apply suggestions from code review: add BrosBboxEmbeddings, replace
einsum with classical matrix operations

jinhopark8345 committed:
Merge remote-tracking branch 'upstream/main' into add-bros

jinhopark8345 commented on src/transformers/models/bros/modeling_bros.py:
```diff
@@ -0,0 +1,1328 @@
+        bbox_pos_emb = bbox_pos_emb.view(seq_length, seq_length, batch_size, d_head)
+
+        bbox_pos_emb = bbox_pos_emb.permute([2, 0, 3, 1])
+        bbox_pos_scores = torch.matmul(
```
@amyeroberts I replaced the `torch.einsum("bnid,bidj->bnij", (query_layer, bbox_pos_emb))` with torch.matmul and and verified that both operations produce the same result. However, after the change, the memory consumption increased significantly. Previously, I could fit a batch_size of 16 with an RTX3090, but now I can only fit a batch_size of 3. Would this be problematic?"

amyeroberts replied:
We definitely want to keep the model usable and not increase memory requirements too much. Let's switch back to the einsum solution. If we find that people are using the model and it's causing issue for them with torchscript then we can work on a memory efficient solution later. Thanks for working on this and verifying the two methods.

jinhopark8345 committed:
Merge remote-tracking branch 'upstream/main' into add-bros

jinhopark8345 committed:
revert einsum update

jinhopark8345 committed:
update bros processor

amyeroberts reviewed (approved):
Great work - thanks for adding this model! 

cc @ArthurZucker for final review 🔍

NielsRogge commented on src/transformers/models/auto/configuration_auto.py:
```diff
@@ -456,6 +458,7 @@
         ("bloom", "BLOOM"),
         ("bort", "BORT"),
         ("bridgetower", "BridgeTower"),
+        ("bros", "Bros"),
```
```suggestion
        ("bros", "BROS"),
```
If you run `make fix-copies` after this, it will update it everywhere.

jinhopark8345 replied:
Applied suggestions from code review!

NielsRogge commented on docs/source/en/_toctree.yml:
```diff
@@ -634,6 +634,8 @@
       - local: model_doc/bridgetower
         title: BridgeTower
+      - local: model_doc/bros
+        title: Bros
```
```suggestion
        title: BROS
```

jinhopark8345 replied:
Applied suggestions from code review!

NielsRogge commented on docs/source/en/model_doc/bros.md:
```diff
@@ -0,0 +1,115 @@
+
+BROS stands for *BERT Relying On Spatiality*. It is an encoder-only Transformer model that takes a sequence of tokens and their bounding boxes as inputs and outputs a sequence of hidden states. BROS encode relative spatial information instead of using absolute spatial information.
+
+It is pre-trained with two objectives: a token-masked LM (TMLM) used in BERT, and a novel area-masked LM (AMLM)
+In TMLM, tokens are randomly masked, and the model predicts the masked tokens using spatial information and other unmasked tokens.
+AMLM is a 2D version of TMLM. It randomly masks text tokens and predicts with the same information as TMLM, but it masks text blocks (areas).
```
```suggestion
It is pre-trained with two objectives: a token-masked language modeling objective (TMLM) used in BERT, and a novel area-masked language modeling objective (AMLM).
In TMLM, tokens are randomly masked, and the model predicts the masked tokens using spatial information and other unmasked tokens.
AMLM is a 2D version of TMLM. It randomly masks text tokens and predicts with the same information as TMLM, but it masks text blocks (areas).
```

jinhopark8345 replied:
Applied suggestions from code review!

jinhopark8345 committed:
apply suggestions from code review

jinhopark8345 committed:
add conversion script for bros

jinhopark8345 committed:
Apply suggestions from code review

jinhopark8345 commented on docs/source/en/model_doc/bros.md:
```diff
@@ -0,0 +1,115 @@
+specific language governing permissions and limitations under the License.
+-->
+
+# Bros
```
Applied suggestions from code review!

jinhopark8345 commented:
@amyeroberts I added `convert_bros_to_pytorch.py` script because `bbox_projection` (linear layer) moved from `BrosTextEmbeddings` to newly added `BrosBboxEmbedding`.

jinhopark8345 committed:
Merge remote-tracking branch 'upstream/main' into add-bros

amyeroberts commented on README.md:
```diff
@@ -310,6 +310,8 @@ Current number of checkpoints: ![](https://img.shields.io/endpoint?url=https://h
 1. **[BLOOM](https://huggingface.co/docs/transformers/model_doc/bloom)** (from BigScience workshop) released by the [BigScience Workshop](https://bigscience.huggingface.co/).
 1. **[BORT](https://huggingface.co/docs/transformers/model_doc/bort)** (from Alexa) released with the paper [Optimal Subarchitecture Extraction For BERT](https://arxiv.org/abs/2010.10499) by Adrian de Wynter and Daniel J. Perry.
 1. **[BridgeTower](https://huggingface.co/docs/transformers/model_doc/bridgetower)** (from Harbin Institute of Technology/Microsoft Research Asia/Intel Labs) released with the paper [BridgeTower: Building Bridges Between Encoders in Vision-Language Representation Learning](https://arxiv.org/abs/2206.08657) by Xiao Xu, Chenfei Wu, Shachar Rosenman, Vasudev Lal, Wanxiang Che, Nan Duan.
+1. **[Bros](https://huggingface.co/docs/transformers/model_doc/bros)** (from NAVER CLOVA) released with the paper [BROS: A Pre-trained Language Model Focusing on Text and Layout for Better Key Information Extraction from Documents](https://arxiv.org/abs/2108.04539) by Teakgyu Hong, Donghyun Kim, Mingi Ji, Wonseok Hwang, Daehyun Nam, Sungrae Park.
+1. **[BROS](https://huggingface.co/docs/transformers/main/model_doc/bros)** (from <FILL INSTITUTION>) released with the paper [<FILL PAPER TITLE>](<FILL ARKIV LINK>) by <FILL AUTHORS>.
```
This needs to be fixed for the documentation to build. Once resolved we can merge 🤗

jinhopark8345 replied:
Applied the fix!

jinhopark8345 committed:
fix readme

jinhopark8345 committed:
apply fix-copies

amyeroberts commented:
@jinhopark8345 Great - thank you! Could you rebase on main to resolve the conflicts? We should be good to go after that :)

jinhopark8345 committed:
Merge remote-tracking branch 'upstream/main' into add-bros

amyeroberts commented:
@jinhopark8345 Thanks for contributing this model! Make sure to share about it's addition to the library on twitter/linkedin/your medium of choice 🤗

ydshieh commented:
@jinhopark8345 Thank you for adding this model into `transformers`.

Regarding the test `BrosModelIntegrationTest.test_inference_no_head`, it fails on our T4 GPU VM as the expected and actual output differ too much, as you can see.

Could you double check on your machines, please? And what's your machine (GPU) type? 

Thank you in advance.

```bash
(Pdb) outputs.last_hidden_state[0, :3, :3]
tensor([[-0.3165,  0.0830, -0.1203],
        [-0.0089,  0.0031,  0.0736],
        [-0.0461,  0.0146,  0.0880]], device='cuda:0')
(Pdb) expected_slice
tensor([[-0.4027,  0.0756, -0.0647],
        [-0.0192, -0.0065,  0.1042],
        [-0.0671,  0.0214,  0.0960]], device='cuda:0')

```

jinhopark8345 commented:
> @jinhopark8345 Thank you for adding this model into `transformers`.
> 
> Regarding the test `BrosModelIntegrationTest.test_inference_no_head`, it fails on our T4 GPU VM as the expected and actual output differ too much, as you can see.
> 
> Could you double check on your machines, please? And what's your machine (GPU) type?
> 
> Thank you in advance.
> 
> ```shell
> (Pdb) outputs.last_hidden_state[0, :3, :3]
> tensor([[-0.3165,  0.0830, -0.1203],
>         [-0.0089,  0.0031,  0.0736],
>         [-0.0461,  0.0146,  0.0880]], device='cuda:0')
> (Pdb) expected_slice
> tensor([[-0.4027,  0.0756, -0.0647],
>         [-0.0192, -0.0065,  0.1042],
>         [-0.0671,  0.0214,  0.0960]], device='cuda:0')
> ```
> 
> You can run the test with
> 
> ```shell
> TF_FORCE_GPU_ALLOW_GROWTH=true RUN_SLOW=1 python3 -m pytest -v tests/models/bros/test_modeling_bros.py::BrosModelIntegrationTest::test_inference_no_head
> ```


@ydshieh Thank you for providing the test command!

I was able to reproduce the issue, but the `outputs.last_hidden_state[0, :3, :3]` value I obtained was different.
Interestingly, not only did the output was different from the `expected_slice` but the value of `outputs.last_hidden_state[0, :3, :3]` changed every time I ran the command you provided. For testing, I am using an RTX3090.

After some testing, I found that some weights weren't being initialized properly.
This issue seems to come from that `bbox_projection` layer (a linear layer) was moved from `BrosEmbeddings` to under `BrosBboxEmbeddings` (`BrosBboxEmbeddings` is newly added class)

By changing:

    model = BrosModel.from_pretrained("naver-clova-ocr/bros-base-uncased").to(torch_device)

to:

    model = BrosModel.from_pretrained("jinho8345/bros-base-uncased").to(torch_device)

I was able to get consistent outputs. (conversion script : `transformers/models/bros/convert_bros_to_pytorch.py`)

I suspect this issue wasn't detected earlier because when running:

    python3 -m pytest -v tests/models/bros/test_modeling_bros.py

torch cuda seed is manually set to certain value, perhaps due to other tests or other reasons.

The update is [here](https://github.com/jinhopark8345/transformers/tree/fix-bros-intergration-test) but I am not sure how I should apply this patch to Transformers library.

ydshieh commented:
Hi @jinhopark8345 Thanks a lot for looking into this!

You can open a PR to update the checkpoint repo used in the test, or we can do it on our own side.

But is it expected that `naver-clova-ocr/bros-base-uncased` doesn't have all the weights? What are the difference between these 2 checkpoints?

jinhopark8345 commented:
> Hi @jinhopark8345 Thanks a lot for looking into this!
> 
> You can open a PR to update the checkpoint repo used in the test, or we can do it on our own side.
> 
> But is it expected that `naver-clova-ocr/bros-base-uncased` doesn't have all the weights? What are the difference between these 2 checkpoints?

 The `naver-clova-ocr/bros-base-uncased` has all the weights. But some weights have been renamed. So if we load `BrosModel` with  `naver-clova-ocr/bros-base-uncased` checkpoint (original checkpoint), the renamed weights won't be initialized correctly with pretrained weights. 
 
 If you confirm updating the checkpoint is okay, I would like to open PR!

ydshieh commented:
Sure, go for it. BTW, I see a lot of `naver-clova-ocr/bros-base-uncased` used, in particular in the examples. So just to be sure, is the user expected to use `naver-clova-ocr/bros-base-uncased` or the renamed one `jinho8345/bros-base-uncased`?

From your description, I think it is `jinho8345/bros-base-uncased`. If this is the case, could you update all occurrence (not just in the tests). Thank you!

ydshieh commented:
Hello @jinhopark8345 Thank you again for fixing the checkpoint. I have yet another question needs your help.

For `BrosModel`, the `bbox_position_embeddings` could be `None` before calling `self.encoder` (if `bbox` is `None`)

https://github.com/huggingface/transformers/blob/37c205eb5d5165b70d3100b599a2bcfc483944f5/src/transformers/models/bros/modeling_bros.py#L927-L933

but eventually, `BrosSelfAttention` will fail if it receive `None` for `bbox_pos_emb`

https://github.com/huggingface/transformers/blob/37c205eb5d5165b70d3100b599a2bcfc483944f5/src/transformers/models/bros/modeling_bros.py#L391

Could you double check if `BrosModel` will only work if `bbox` is not `None` in the original implementation? If this is not the case, how is `bbox_pos_emb` being created if `bbox` is `None` etc.

Thank you in advance, again!

jinhopark8345 commented:
Hello @ydshieh Thank you for asking!

Below code is the original implementation
``` 
        scaled_bbox = bbox * self.config.bbox_scale
        bbox_pos_emb = self.embeddings.calc_bbox_pos_emb(
            scaled_bbox, self.config.pe_type
        )
```

In original implementation, `BrosModel` will only work if `bbox` is not  `None`.

Would it be more helpful to users if we remove https://github.com/huggingface/transformers/blob/2d71307dc0ee2849f785568f345837e726209fc6/src/transformers/models/bros/modeling_bros.py#L928 so that `BrosModel` fails earlier? or do you suggest different solutions?

ydshieh commented:
Hi! In this case, you can add a try: except at the beginning of `BrosModel.forward` method as the input validation.

(we might need a few more fixes if CI fails due to this)

Thank you !

NielsRogge commented:
Hi @jinhopark8345 , congrats on this amazing contribution.

Feel free to share about it on Twitter/LinkedIn and we'll amplify.

Prathyusha-Akundi commented:
Hi @jinhopark8345,
Can you please provide examples of how to use logits from BrosSpadeELForTokenClassification to identify the intra-relationships? 
TIA

jinhopark8345 commented:
Hi @Prathyusha-Akundi,

You can refer to the [example notebook](https://github.com/jinhopark8345/FormUnderstanding/blob/main/notebooks/Fine_tuning_bros_spade_on_FUNSD_entity_linking_dataset.ipynb) for identifying intra-relationships.

If you are looking for information on entity linking versus entity extraction, you can check out the [entity linking explanation vs entity extraction](https://github.com/jinhopark8345/FormUnderstanding) here.

Prathyusha-Akundi commented:
Thank you @jinhopark8345 , this is extremely helpful!
